dm/pkg/checker/table_structure.go

```diff
@@ -296,168 +343,140 @@ func (c *ShardingTablesChecker) Check(ctx context.Context) *Result {
-					if isMySQLError(err, mysql.ErrNoSuchTable) {
-						continue
-					}
-					markCheckError(r, err)
-					r.Extra = fmt.Sprintf("instance %s on sharding %s", instance, c.name)
-					return r
-				}
+	startTime := time.Now()
+	log.L().Logger.Info("start to check sharding tables")
 
-				info, err := dbutil.GetTableInfoBySQL(statement, parser2)
-				if err != nil {
-					markCheckError(r, err)
-					r.Extra = fmt.Sprintf("instance %s on sharding %s", instance, c.name)
-					return r
-				}
-				stmt, err := parser2.ParseOneStmt(statement, "", "")
-				if err != nil {
-					markCheckError(r, errors.Annotatef(err, "statement %s", statement))
-					r.Extra = fmt.Sprintf("instance %s on sharding %s", instance, c.name)
-					return r
-				}
+	for sourceID, tables := range c.tableMap {
+		c.firstSourceID = sourceID
+		c.firstTable = tables[0]
+		break
+	}
+	db, ok := c.dbs[c.firstSourceID]
+	if !ok {
+		markCheckError(r, errors.NotFoundf("client for sourceID %s", c.firstSourceID))
+		return r
+	}
 
-				ctStmt, ok := stmt.(*ast.CreateTableStmt)
-				if !ok {
-					markCheckError(r, errors.Errorf("Expect CreateTableStmt but got %T", stmt))
-					r.Extra = fmt.Sprintf("instance %s on sharding %s", instance, c.name)
-					return r
-				}
+	p, err := dbutil.GetParserForDB(ctx, db)
+	if err != nil {
+		r.Extra = fmt.Sprintf("fail to get parser for sourceID %s on sharding %s", c.firstSourceID, c.targetTableID)
+		markCheckError(r, err)
+		return r
+	}
+	r.Extra = fmt.Sprintf("sourceID %s on sharding %s", c.firstSourceID, c.targetTableID)
+	statement, err := dbutil.GetCreateTableSQL(ctx, db, c.firstTable.Schema, c.firstTable.Name)
+	if err != nil {
+		markCheckError(r, err)
+		return r
+	}
 
-				if c.checkAutoIncrementPrimaryKey {
-					passed := c.checkAutoIncrementKey(instance, schema, table, ctStmt, info, r)
-					if !passed {
-						return r
-					}
-				}
+	c.firstCreateTableStmtNode, err = getCreateTableStmt(p, statement)
+	if err != nil {
+		markCheckError(r, err)
+		return r
+	}
 
-				if stmtNode == nil {
-					stmtNode = ctStmt
-					firstTable = dbutil.TableName(schema, table)
-					firstInstance = instance
-					continue
-				}
+	concurrency, err := getConcurrency(ctx, c.tableMap, c.dbs, c.dumpThreads)
+	if err != nil {
+		markCheckError(r, err)
+		return r
+	}
+	eg, checkCtx := errgroup.WithContext(ctx)
+	for i := 0; i < concurrency; i++ {
+		eg.Go(func() error {
+			return c.checkShardingTable(checkCtx, r)
+		})
+	}
 
-				checkErr := c.checkConsistency(stmtNode, ctStmt, firstTable, dbutil.TableName(schema, table), firstInstance, instance)
-				if checkErr != nil {
-					r.State = StateFailure
-					r.Errors = append(r.Errors, checkErr)
-					r.Extra = fmt.Sprintf("error on sharding %s", c.name)
-					r.Instruction = "please set same table structure for sharding tables"
-					return r
-				}
-			}
-		}
+	dispatchTableItem(checkCtx, c.tableMap, c.inCh)
+	if err := eg.Wait(); err != nil {
+		c.reMu.Lock()
```

Comment:
Why sometimes lock but sometimes not? e.g. L384

Reply:
only one thread uses `r` before `eg.Go()`

Reply:
After eg.Wait(), all checkShardingTable exit.

Reply:
how about add lock inner func `markCheckError`? i mean add a mutex into `Result` struct, beacuse `reMu` is just used to lock `result`

Reply:
GMH means there is also one thread after eg.Wait(), so we needn't lock() them after eg.Wait(). @Ehco1996

Reply:
yes, but this relay on current logic to make sure no data-race hanppened, i am worried that something might go wrong here when later refactor.

there are many thread trying to wirte result object before `eg.Wait()`, so we still need a lock . if you move lock/unlock operations into `markCheckError` there is no querstion about `Why sometimes lock but sometimes not? e.g. L384` any more